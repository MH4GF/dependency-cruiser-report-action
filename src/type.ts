import { getOctokit } from '@actions/github'

export type Octokit = ReturnType<typeof getOctokit>
