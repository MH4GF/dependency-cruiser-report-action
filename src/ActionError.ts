import * as core from '@actions/core'
type Severity = 'warning' | 'error'

export class ActionError extends Error {
  readonly severity: Severity

  constructor(message: string, severity: Severity) {
    super(message)
    this.severity = severity
  }
}

export const exitWithMessage = (error: ActionError) => {
  switch (error.severity) {
    case 'warning':
      process.exitCode = core.ExitCode.Success
      core.warning(error.message)
      break
    default:
      core.setFailed(error.message)
  }
}
