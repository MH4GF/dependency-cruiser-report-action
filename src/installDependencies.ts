import { exec } from '@actions/exec'

export const installDependencies = async () => {
  return await exec('yarn install')
}
