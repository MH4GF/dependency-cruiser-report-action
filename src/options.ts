import { existsSync } from 'fs'

import * as core from '@actions/core'
import { context } from '@actions/github'

import { filterSupportedFiles } from './options/filterSupportedFiles'
import { formatFocusOption } from './options/formatFocusOption'
import { Options, validateOptions } from './options/validateOptions'

const mayBeConfigFilePath = () => {
  const path = '.dependency-cruiser.js'
  return existsSync(path) ? path : ''
}

const getConfigFilePath = () => {
  const depcruiseConfigFile = core.getInput('config_file', { required: false })
  return depcruiseConfigFile !== '' ? depcruiseConfigFile : mayBeConfigFilePath()
}

const getSha = (): string =>
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-member-access
  context.payload.after ?? context.payload.pull_request?.head?.sha

export const getOptions = (): Promise<Options> => {
  const token = core.getInput('github_token', { required: true })
  const workingDirectory = core.getInput('working_directory', { required: true })
  const changedFiles = core.getInput('target_files', { required: false }).split(' ')
  const targetFiles = filterSupportedFiles(changedFiles)
  const focus = formatFocusOption(targetFiles)
  const cruiseScript = core.getInput('cruise_script', { required: true })
  const depcruiseConfigFilePath = getConfigFilePath()
  const pr = context.payload.pull_request
  const options = {
    token,
    workingDirectory,
    owner: context.repo.owner,
    repo: context.repo.repo,
    issueNumber: pr?.number,
    sha: getSha(),
    targetFiles: targetFiles.join(' '),
    focus,
    depcruiseConfigFilePath,
    cruiseScript,
  }

  return validateOptions(options)
}

export type { Options } from './options/validateOptions'
