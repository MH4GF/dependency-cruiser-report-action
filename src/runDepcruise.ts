import { exec } from '@actions/exec'

type Options = {
  targetFiles: string
  focus: string
  depcruiseConfigFilePath: string
  cruiseScript: string
}

type DepcruiseResult = {
  cmdText: string
  mermaidText: string
}

export const runDepcruise = async ({
  targetFiles,
  focus,
  depcruiseConfigFilePath,
  cruiseScript,
}: Options): Promise<DepcruiseResult> => {
  const outputTypeOption = '--output-type mermaid'
  const configOption = depcruiseConfigFilePath !== '' ? `--config ${depcruiseConfigFilePath}` : ''
  const focusOption = `--focus ${focus}`
  const cmd = `${cruiseScript} ${outputTypeOption} ${configOption} ${focusOption} ${targetFiles}`
  const options = { listeners: {} }
  let mermaid = ''
  options.listeners = {
    stdout: (data: Buffer) => {
      mermaid += data.toString()
    },
  }
  await exec(cmd, [], options)

  return { mermaidText: mermaid, cmdText: cmd }
}
