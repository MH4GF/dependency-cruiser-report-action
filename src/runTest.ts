import { exec } from '@actions/exec'

export const runTest = async (): Promise<number> => {
  return await exec(
    'npx -p dependency-cruiser depcruise --include-only "^src" --output-type dot src | dot -T svg > dependencygraph.svg',
  )
}
