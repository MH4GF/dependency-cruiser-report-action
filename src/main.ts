import * as core from '@actions/core'
import { getOctokit } from '@actions/github'

import { ActionError, exitWithMessage } from './ActionError'
import { installDependencies } from './installDependencies'
import { getOptions } from './options'
import { generateReport } from './report/generateReport'
import { runDepcruise } from './runDepcruise'

export const run = async (): Promise<void> => {
  const options = await getOptions()
  const octokit = getOctokit(options.token)

  await installDependencies()
  const { mermaidText, cmdText } = await runDepcruise(options)
  await generateReport(octokit, options, mermaidText, cmdText)
}

run().catch((error) => {
  if (error instanceof ActionError) exitWithMessage(error)
  else if (error instanceof Error) core.setFailed(error.message)
})
