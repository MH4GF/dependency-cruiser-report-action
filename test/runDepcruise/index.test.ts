import { describe, expect, it } from 'vitest'

import { runDepcruise } from '../../src/runDepcruise'

const defaultOptions = {
  targetFiles: 'test/runDepcruise/sample/__mocks__/test',
  focusFiles: `"^test/runDepcruise/sample/__mocks__/test/fixtures/cjs/root_one.js|^test/runDepcruise/sample/__mocks__/test/fixtures/cjs/root_two.js"`,
  depcruiseConfigFilePath: 'test/runDepcruise/.dependency-cruiser.js',
  cruiseScript: 'pnpm exec depcruise',
}

describe('runDepcruise', () => {
  it('execute depcruise command(--focus)', async () => {
    const options = { ...defaultOptions, visualizeType: 'focus' }
    const result = await runDepcruise(options)

    expect(result.mermaidText).toMatchSnapshot()
  })
  it('execute depcruise command(--reaches)', async () => {
    const options = { ...defaultOptions, visualizeType: 'reaches' }
    const result = await runDepcruise(options)

    expect(result.mermaidText).toMatchSnapshot()
  })
})
