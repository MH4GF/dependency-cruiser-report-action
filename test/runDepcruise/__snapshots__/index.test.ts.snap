// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runDepcruise > execute depcruise command(--focus) 1`] = `
"flowchart LR

subgraph 0["test"]
subgraph 1["runDepcruise"]
subgraph 2["sample"]
subgraph 3["__mocks__"]
subgraph 4["test"]
subgraph 5["fixtures"]
subgraph 6["cjs"]
7["one_only_two.js"]
8["root_one.js"]
9["shared.js"]
subgraph A["sub"]
B["dir.js"]
end
C["root_two.js"]
D["somedata.json"]
E["two_only_one.js"]
end
end
end
end
end
end
end
8-->7
8-->9
8-->B
C-->9
C-->D
C-->E
E-->B

style 8 fill:lime,color:black
style C fill:lime,color:black"
`;

exports[`runDepcruise > execute depcruise command(--reaches) 1`] = `
"flowchart LR

subgraph 0["test"]
subgraph 1["runDepcruise"]
subgraph 2["sample"]
subgraph 3["__mocks__"]
subgraph 4["test"]
subgraph 5["fixtures"]
subgraph 6["cjs"]
7["root_one.js"]
8["root_two.js"]
end
end
end
end
end
end
end


style 7 fill:lime,color:black
style 8 fill:lime,color:black"
`;
